// @springworks/error-factory/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.create = undefined;
exports.createError = createError;

var _util = require('util');

var _fel = require('fel');

var fel = _interopRequireWildcard(_fel);

var _internals = require('./internals');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function createError(opt_params) {
  const params = opt_params || {};
  if (!params.stack_from) {
    params.stack_from = createError;
  }
  const error = fel.createError(params);
  Object.defineProperties(error, {
    toJSON: { value: _internals.getOutputPayload, enumerable: true },
    payload: (0, _internals.createLogPayloadDescriptor)(),
    statusCode: _internals.deprecatedStatusCodeProperty,
    body: _internals.deprecatedBodyProperty,
    json: _internals.deprecatedJsonProperty
  });
  return error;
}

const create = exports.create = (0, _util.deprecate)(legacyCreate, 'error-factory create is deprecated. Use createError');

function legacyCreate() {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  const params = (0, _internals.patternmatchLegacyArgs)(args);
  params.stack_from = create;
  return createError(params);
}
